[![Python 3.10 | 3.12](https://img.shields.io/badge/python-3.10%20%7C%203.12-blue)](https://www.python.org/downloads/)

# rnalib: a Python-based transcriptomics library
[<img alt="rnalib" src="docs/_static/rnalib_logo.png" align="right" width="200"/>](https://rnalib.readthedocs.io/en/latest/)


Rnalib is a python library for handling transcriptomics data.

## Installation

To install the latest release of rnalib, use the following command:

```bash
pip install rnalib
```

To install the latest version of rnalib from source, use the following command:

```bash
pip install git+https://github.com/popitsch/rnalib.git
```

## Quickstart
![API demo](docs/_static/screencasts/introduction.gif)

Please see the [rnalib's documentation](https://rnalib.readthedocs.io/en/latest/) for more information.

[<img alt="readthedocs" src="https://about.readthedocs.com/theme/img/logo-wordmark-dark.svg" width="300"/>](https://rnalib.readthedocs.io/en/latest/)

## Notebooks

We provide a set of ipython notebooks for demonstrating *rnalib*'s API:
* [README.ipynb](notebooks/README.ipynb): basic introduction to *rnalib*
* [AdvancedUsage.ipynb](notebooks/AdvancedUsage.ipynb): advanced API usage

We also provide a (growing) set of tutorials for demonstrating *rnalib* in real-world use cases:
* [Tutorial: SLAM-seq time-course data analysis](notebooks/Tutorial_SLAM-seq.ipynb)
* [Tutorial: Comparison of different gene annotation sets (human and fly)](notebooks/Tutorial_compare_annotation_sets.ipynb)
* [Tutorial: Transcriptome annotation with genemunge, archs4 and mygene.info](notebooks/Tutorial_transcriptome_annotation.ipynb)
* [Tutorial: CTCF analysis with rnalib and bioframe](notebooks/Tutorial_CTCF_analysis.ipynb)
* [Tutorial: A small analysis of shRNA targets](notebooks/Tutorial_shRNA_analysis.ipynb)

* We also compare *rnalib* to other genomics libraries wrt. performance in this [RelatedWork](notebooks/RelatedWork_performance.ipynb) notebook.
